### PR TITLE
fix #5118 Dual Mesh Node Extended Algorithm

### LINKED ISSUE #5118: Dual Mesh Node Extended Algorithm
Sverchok 1.2-Alpha

image

1. Geometry Node Dual Mesh has property "Keep Boundaries"
image

2. Original Sverchok Dual Mesh node has some glitches if geometry is non-manifold:

image
image

3. Geometry Nodes->Dual Mesh has glitches too:
image

Lets update DualMesh algorithm?

### PR DESCRIPTION
**Pull request by satabol**
fix #5118
- move DualMesh to old_nodes/dual_mesh.py
- create SvDualMeshNodeMK2

test file:
[Issue.DualMesh.Tests.003.blend.zip](https://github.com/nortikin/sverchok/files/15214178/Issue.DualMesh.Tests.003.blend.zip)

Examples in test file:

image

### CODE REVIEW

**Comment by satabol** [comment]
image

**Comment by satabol** [comment]
Added new input socket "Levels":
DualMesh levels 001

**Comment by satabol** [comment]
Second Level issue:

image

**Comment by satabol** [comment]
Second Level Issue:
image

**Comment by satabol** [comment]
Images for comment in code

image

image
image
image
image
image
image
image

File: nodes/spatial/mesh_clustering.py [hunk 1]
```diff
@@ -232,7 +232,7 @@ def process(self):
             # if(cluster_subdivide_i>0):
             #     clust.subdivide(cluster_subdivide_i)
             
-            clust.cluster(cluster_counts_i, maxiter=max_iter_i)
+            clust.cluster(cluster_counts_i, maxiter=max_iter_i) # https://github.com/pyvista/pyacvd/blob/6cdecab61bf566bb9ebee197c5294bf607cb936d/pyacvd/cython/_clustering.pyx#L94
```
Comment (portnov) [inline nodes/spatial/mesh_clustering.py:235]: this change probably should go to another PR? :)

File: utils/sv_bmesh_utils.py [hunk 2]
```diff
@@ -471,45 +472,426 @@ def remove_doubles(vertices, edges, faces, d, face_data=None, vert_data=None, ed
...
 
 
-def dual_mesh(bm, recalc_normals=True):
-    # Make vertices of dual mesh by finding
-    # centers of original mesh faces.
-    new_verts = dict()
-    for face in bm.faces:
-        new_verts[face.index] = face.calc_center_median()
+def dual_mesh(bm, recalc_normals=True, keep_boundaries=False):
+    from time import time_ns
```
Comment (portnov) [inline utils/sv_bmesh_utils.py:476]: is it still needed?

File: utils/sv_bmesh_utils.py [hunk 3]
```diff
@@ -471,45 +472,426 @@ def remove_doubles(vertices, edges, faces, d, face_data=None, vert_data=None, ed
...
 
-    # For each vertex of original mesh,
-    # find all connected faces and connect
-    # corresponding vertices of the dual mesh
-    # with a face.
-    # The problem is, that the order of edges in
-    # vert.link_edges (or faces in vert.link_faces)
-    # is undefined, so we have to sort them somehow.
+    # some definitions:
+    # 1. source vertices, edges, faces - elements of cource mesh. Some vertices, edges and faces may be as single.
```
Comment (portnov) [inline utils/sv_bmesh_utils.py:482]: *source.

File: nodes/spatial/mesh_clustering.py (hunk 1 above)
Comment (satabol) [reply nodes/spatial/mesh_clustering.py:235]: It can be removed. I will. I try to refactor and set this link to memory.

File: utils/sv_bmesh_utils.py (hunk 2 above)
Comment (satabol) [reply utils/sv_bmesh_utils.py:476]: For a little time

File: utils/sv_bmesh_utils.py (hunk 3 above)
Comment (satabol) [reply utils/sv_bmesh_utils.py:482]: fixed
